src/mscorlib/src/System/String.cs

```diff
@@ -159,27 +159,26 @@ public sealed class String : IComparable, ICloneable, IConvertible, IEnumerable
             Contract.Ensures(Contract.Result<String>() != null);
             Contract.EndContractBlock();
 
-            if (separator == null)
-                separator = String.Empty;
-
-
             using(IEnumerator<String> en = values.GetEnumerator()) {
                 if (!en.MoveNext())
                     return String.Empty;
 
-                StringBuilder result = StringBuilderCache.Acquire();
-                if (en.Current != null) {
-                    result.Append(en.Current);
-                }
+                String value = en.Current;
 
-                while (en.MoveNext()) {
-                    result.Append(separator);
-                    if (en.Current != null) {
+                if (en.MoveNext()) { 
```

Comment:
The `en.Current` value needs to be captured before the `en.MoveNext()` because afterwards it will be on the second row's data (or null if finished).